### ISSUE TITLE
Migrate away from tap in all repositories

### ISSUE THREAD

**Issue by mcollina** [fastify/fastify#5555]
I think we __must__ move away from
tap in favor of node:test, as Isaacs has stated that he will drop unsupported Node.js version in patch releases. Given our infrequent major release cadence, it's not a good bet for us.

https://x.com/izs/status/1810339751456825767?s=46&t=22eoAstJVk5l46KQXYEk5Q

**Comment by jsumners** [comment]
100% on board. Whatever my quibbles with `node:test` may be, the new position of `tap` is untenable.

**Comment by metcoder95** [comment]
+1

**Comment by climba03003** [comment]
Can we delay this work until `fastify@5` is released as latest?
Currently, most new PR is landed in `main` then back-ported to `4.x`. Such changes would break the automatic back-porting immediately.

**Comment by mcollina** [comment]
I think so!

**Comment by jsumners** [comment]
Yes. Dev dependencies do not affect releases.

**Comment by marios-codes** [comment]
Hello! Sorry if this a dumb answer. Regarding the issue does someone need to also rewrite all the test cases to use node:test?

**Comment by Uzlopak** [comment]
The needed rewrite is usually trivial.

**Comment by jsumners** [comment]
Feel free to jump in and work on this task!

We've decided not to assign the issue to avoid any ["cookie licking"](https://www.redhat.com/en/blog/dont-lick-cookie), so please remember these simple guidelines:

- First, check if there are any existing Pull Requests (PRs) related to this issue. Just check the notification in this page.

- If you create a PR or draft PR, make sure to link it to this issue using the [GitHub keyword method](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword). It helps keep things organized, and you'll see a nice message here.

Don't worry about whether this issue is assigned to your GitHub handle or not—just dive right in. We're all about collaboration here!

**Comment by marios-codes** [comment]
Thank you very much🙂

**Comment by Uzlopak** [comment]
Just for you to get an idea of the "rewrite" part.

before
```js
'use strict'

const { test } = require('tap')
const { isAsyncFunction } = require('../../lib/is-async-function.js')

test('isAsyncFunction returns true for async functions', (t) => {
  t.plan(1)
  t.ok(isAsyncFunction(async () => { }))
})

test('isAsyncFunction returns false for non-async functions', (t) => {
  t.plan(1)
  t.notOk(isAsyncFunction(() => { }))
})

```

after:
```js
'use strict'

const { test } = require('node:test')
const { isAsyncFunction } = require('../../lib/is-async-function.js')

test('isAsyncFunction returns true for async functions', (t) => {
  t.plan(1)
  t.assert.ok(isAsyncFunction(async () => { }))
})

test('isAsyncFunction returns false for non-async functions', (t) => {
  t.plan(1)
  t.assert.ok(!isAsyncFunction(() => { }))
})

```

**Comment by marios-codes** [comment]
Which node version does main branch (v5) support as I cannot find it in the usual places? Thank you!

**Comment by marios-codes** [comment]
Can you please elaborate a bit more as I am totally new to contributing? I mean should I work on the main (on a different branch of course) or should I work on v4 regarding the migration of the unit tests?

**Comment by Uzlopak** [comment]
https://github.com/search?q=topic%3Afastify-plugin+org%3Afastify&type=Repositories

**Comment by Uzlopak** [comment]
I think a good blueprint for this task is the migration made by @dancastillo for @fastify/routes

https://github.com/fastify/fastify-routes/pull/107

**Comment by nimesh0505** [comment]
Hello @mcollina,

I'm working on this issue. While starting this task for the [pre-commit](https://github.com/fastify/pre-commit) repo, I realized we should first address the Node.js version support.

Given that Node.js 16 has reached its end of life, I propose the following steps before proceeding with the test framework migration:

1. Remove Node.js 16 from the CI configuration in `.github/workflows/ci.yml`.
   ```
   node-versions: '["18", "20", "22"]'
   ```
2. Update the `package.json` to reflect the new minimum supported Node.js version:
   ```json
   "engines": {
     "node": ">=18"
   }
   ```

This change aligns with the current LTS versions of Node.js and ensures we're not supporting EOL versions.

After making these changes, I'll proceed with the migration from tap to node:test.

What are your thoughts on this approach? 
Thank you for your guidance!

**Comment by mcollina** [comment]
We are targeting v20+ for all the new releases of Fastify, go ahead with that in pre-commit too.

**Comment by narendra-reddy-nf** [comment]
Hi @mcollina,

I will be working on migrating the tests from `tap` to `node:test` in the [fastify-flash](https://github.com/fastify/fastify-flash) repository, using Node.js version v20 as the baseline, as suggested.

Let me know if you have any further instructions!

**Comment by narendra-reddy-nf** [comment]
Hey @mcollina,

I’m migrating the fastify-flash repo to use node:test and could use your advice on a few things:

1. TypeScript Compatibility: Since node:test doesn’t support TypeScript by default, I’ve been exploring options. I’ve found that tsx and ts-node can run TypeScript tests without precompiling, which is helpful. I’m also looking into test loaders but still need to test this setup.
2. Test Reporting: Right now, the project uses tap-mocha-reporter, and I saw that node:test has a built-in JUnit reporter. I’m not sure how well it will work with TypeScript tests, so I may need to experiment to figure that out.
3. Considering borp: I found a test runner called borp, which works with TypeScript and uses node:test. It seems like a good option since it also handles coverage with c8 and provides reporting tools. I’m wondering if switching to it might make things easier overall.

I’d appreciate any thoughts you have on the best way to handle this, especially around TypeScript support and reporting.

Thanks!

**Comment by mcollina** [comment]
I actually `borp` for that specific reason. Also using `tsx` would work very well for this.

**Comment by reidn3r** [comment]
Can i work on this issue too? Maybe by refactoring a module?

**Comment by gurgunday** [comment]
Yeah! Everyone’s contributions are welcome!

**Comment by atharva123a** [comment]
hello folks! This looked like a good first issue, so i decided to clone the repo and run tests to start off just to make sure everything is working alright.

But there are some issues that are showing up when I run npm run test:

Some files are failing tests: 
`FAILED: fastify/test/http-methods/proppatch.test.js`

and I get this warning:

> Warning:: Test "Encapsulation works" generated asynchronous activity after the test ended. This activity created the error "TypeError: Cannot read properties of undefined (reading 'strictEqual')" and would have caused the test to fail, but instead triggered an uncaughtException event.
Warning:: Test "Encapsulation works" generated asynchronous activity after the test ended. This activity created the error "FastifyError [FST_ERR_PLUGIN_TIMEOUT]: fastify-plugin: Plugin did not start in time: '(instance, opts, done) => { -- instance.decorate('use', () => true)'. You may have forgotten to call 'done' function or to resolve a Promise" and would have caused the test to fail, but instead triggered an unhandledRejection event.

I am currently using node version: **v20.10.0**

The repo is up to date with the main branch. 

I want to work on this and could really use any help/guidance related to this!

**Comment by climba03003** [comment]
> I want to work on this and could really use any help/guidance related to this!

Please update your node version to at least `20.15.0`.
I recommend to always use the latest LTS version, since we only guarantee to works on latest version of each major.

**Comment by SamSalvatico** [comment]
Is anybody working on this in relation to [Fastify Autoload](https://github.com/fastify/fastify-autoload)?
I've seen that there we have mixed vitest/tap approaches, maybe I could work on this step by step
I cannot see any related PR, but just want to ensure

**Comment by Uzlopak** [comment]
Back in those days vitest was the best solution. For the sake of uniformity I would also recommend to move to node test runner.

**Comment by mcollina** [comment]
> Is anybody working on this in relation to [Fastify Autoload](https://github.com/fastify/fastify-autoload)?
I've seen that there we have mixed vitest/tap approaches, maybe I could work on this step by step
I cannot see any related PR, but just want to ensure

Basically we need to support all those test frameworks because they do quite a lot of funky things with the Node.js process and transpilation. All tap tests should move to `node:test`, the rest should stay as they are.

**Comment by SamSalvatico** [comment]
> > Is anybody working on this in relation to [Fastify Autoload](https://github.com/fastify/fastify-autoload)?
> > I've seen that there we have mixed vitest/tap approaches, maybe I could work on this step by step
> > I cannot see any related PR, but just want to ensure
> 
> Basically we need to support all those test frameworks because they do quite a lot of funky things with the Node.js process and transpilation. All tap tests should move to `node:test`, the rest should stay as they are.

Digging into it I noticed the need of the vitest coverage, thanks 
I will work on the tap ones

**Comment by IcaroSilvaFK** [comment]
```js 
test('async hooks', (t, done) => {
  t.plan(21)
  const fastify = Fastify({ exposeHeadRoutes: false })
  fastify.addHook('onRequest', async function (request, reply) {
    await sleep(1)
    request.test = 'the request is coming'
    reply.test = 'the reply has come'
    if (request.raw.method === 'DELETE') {
      throw new Error('some error')
    }
  })

  fastify.addHook('preHandler', async function (request, reply) {
    await sleep(1)
    t.assert.equal(request.test, 'the request is coming')
    t.assert.equal(reply.test, 'the reply has come')
    if (request.raw.method === 'HEAD') {
      throw new Error('some error')
    }
  })

  fastify.addHook('onSend', async function (request, reply, payload) {
    await sleep(1)
    t.assert.ok('onSend called')
  })

  fastify.addHook('onResponse', async function (request, reply) {
    await sleep(1)
    t.assert.ok('onResponse called')
  })

  fastify.get('/', function (request, reply) {
    t.assert.equal(request.test, 'the request is coming')
    t.assert.equal(reply.test, 'the reply has come')
    reply.code(200).send({ hello: 'world' })
  })

  fastify.head('/', function (req, reply) {
    reply.code(200).send({ hello: 'world' })
  })

  fastify.delete('/', function (req, reply) {
    reply.code(200).send({ hello: 'world' })
  })

  fastify.listen({ port: 0 }, async err => {
    t.assert.ifError(err)
    t.after(() => { 
      fastify.close() 
    })

    sget({
      method: 'GET',
      url: getServerUrl(fastify)
    }, (err, response, body) => {
      t.assert.ifError(err)
      t.assert.equal(response.statusCode, 200)
      t.assert.equal(response.headers['content-length'], '' + body.length)
      t.assert.deepStrictEqual(JSON.parse(body), { hello: 'world' })
    })

    sget({
      method: 'HEAD',
      url: getServerUrl(fastify)
    }, (err, response, body) => {
      t.assert.ifError(err)
      t.assert.equal(response.statusCode, 500)
    })

    sget({
      method: 'DELETE',
      url: getServerUrl(fastify)
    }, (err, response, body) => {
      t.assert.ifError(err)
      t.assert.equal(response.statusCode, 500)
      done()
    })
  })
})

```
I have a test that makes 21 calls using “execution waiting”, but when migrating to node:test it does not wait for all callbacks and ends up returning almost random values. Even using the waitForCb function, the execution does not pause correctly before moving on. It would be interesting to encapsulate this logic in a Promise to ensure that the test only progresses when all callbacks have completed.
In this example
```js
await new Promise((resolve) => {
    sget({
      method: 'GET',
      url: getServerUrl(fastify)
    }, (err, response, body) => {
      t.assert.ifError(err)
      t.assert.equal(response.statusCode, 200)
      t.assert.equal(response.headers['content-length'], '' + body.length)
      t.assert.deepStrictEqual(JSON.parse(body), { hello: 'world' })  

      resolve()
    })
  })

// or

const {promisify} = require("node:util")
const sgetPromisse = promisify(sget)

```

**Comment by avifenesh** [comment]
@mcollina @gurgunday @ofek29
Is there a place where the tests needing migration are tracked, to know where contribution is needed?

**Comment by jsumners** [comment]
@IcaroSilvaFK what is your question?

**Comment by jsumners** [comment]
@avifenesh no, this is the only tracking issue we have. You'll need to grep for `tap` and make sure the one you pick doesn't already have an open PR.

**Comment by avifenesh** [comment]
> @avifenesh no, this is the only tracking issue we have. You'll need to grep for `tap` and make sure the one you pick doesn't already have an open PR.

@ofek23

**Comment by Deepcoders30** [comment]
Hi! I’d like to work on this issue as my first contribution. Could you please assign it to me? Thanks!

**Comment by jsumners** [comment]
Feel free to jump in and work on this task!

We've decided not to assign the issue to avoid any ["cookie licking"](https://www.redhat.com/en/blog/dont-lick-cookie), so please remember these simple guidelines:

- First, check if there are any existing Pull Requests (PRs) related to this issue. Just check the notification in this page.

- If you create a PR or draft PR, make sure to link it to this issue using the [GitHub keyword method](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword). It helps keep things organized, and you'll see a nice message here.

Don't worry about whether this issue is assigned to your GitHub handle or not—just dive right in. We're all about collaboration here!

**Comment by satoharuto961-dev** [comment]
I agree with the reasoning overall. You’ve clearly linked the recommendation to Node.js version support and our release cadence, which makes the argument concrete. A couple of points that could make it even stronger:

Clarify the timeline: It might help to specify which Node.js versions are affected and when Isaacs plans to drop support, so readers can quickly see the urgency.
Suggest a migration path: Mentioning any potential steps or considerations for moving from tap to node:test would make the recommendation more actionable.
Tone adjustment (optional): “It’s not a good bet for us” is clear but slightly informal. You could rephrase as “This introduces risk for our release schedule,” which keeps it professional and precise.

Overall, it’s a solid technical point with good reasoning; just adding a bit more context and actionable guidance would make it even stronger.

**Comment by zain-asif-dev** [comment]
Migrated `fastify/fastify-template-gitpod` off tap, since it was the last repo I could find in the org still depending on it (checked via GitHub code search for `require('tap')` across `org:fastify`, plus manually checking `package.json` in the fastify-template-gitpod repo and several plugin repos, most of which already appear to be on `node:test`).

PR: https://github.com/fastify/fastify-template-gitpod/pull/10
